### ISSUE TITLE
Reused state crashes subsequent backward passes

### ISSUE THREAD

**Issue by Jegp** [norse/norse#130]
We're currently asking the state voltage to employ gradients for later backward passes. That works if the state is renewed for every cycle, but this crashes: 

```python

    model = LIFCell(12, 1)
    data = torch.ones(100, 12)
    out, s = model(data)
    loss = out.sum()
    loss.backward()
    out, _ = model(data, s)
    loss = out.sum()
```

**Comment by cpehle** [comment]
I don't think this is really a bug. You can't reuse variables in multiple backward passes without retaining the graph or detaching the variable.

This works:
```python
model = LIFCell(12, 1)
data = torch.ones(100, 12)
s = None

for ts in range(10):
    out, s = model(data, s)

loss = out.sum()
loss.backward()
```
and this is the use case that the module is meant for (and this use case would be broken by the fix that you propose).  If for some reason the use wanted to do what you consider above he would have to detach s from the previous computation graph himself. It could be seen as pytorch bug that it outright crashes, but on the other hand this is an unintended use case.

**Comment by cpehle** [comment]
By the way, this:

```python
import norse
import torch

model = norse.LIFCell(12, 1)
data = torch.ones(100, 12)

out, s = model(data)
loss = out.sum()
loss.backward(retain_graph=True)
out, _ = model(data, s)
loss = out.sum()
```
also works.

**Comment by Jegp** [comment]
I completely agree on that. It's never the point that this should be allowed. What I was imagining we could do is to take care of this gradient handling for the blissfully ignorant user. 
I think I was trying to be a bit too clever. I don't think there's any nice way of getting around this, and I think we should probably close and forget this.

It's also the convention of the stateful torch modules to throw on the same use case.

**Comment by cpehle** [comment]
> I completely agree on that. It's never the point that this should be allowed. What I was imagining we could do is to take care of this gradient handling for the blissfully ignorant user.
> I think I was trying to be a bit too clever. I don't think there's any nice way of getting around this, and I think we should probably close and forget this.

Glad we agree :). This definitely is a sharp edge of the PyTorch API and is particularly hard to get right for the issues (#87, #105) related to STDP and other rules that operate on synaptic weights.

**Comment by Jegp** [comment]
Indeed. And I think we have a bit of documentation to do. Especially because we're so dependent on BPTT and other niceties.

Let's see how this unravels in the future.
